_posts/2021-03-29-apply-pretrained-pytorch-model.md

```diff
@@ -0,0 +1,154 @@
+---
+layout: post
+title: Dask with PyTorch for large scale image analysis
+author: Nicholas Sofroniew, Matthew Rocklin
+tags: [imaging]
+theme: twitter
+---
+{% include JB/setup %}
+
+Executive Summary
+-----------------
+
+This post explores applying a pre-trained [PyTorch](https://pytorch.org/) model in parallel with Dask Array.
+
+We cover a simple example applying a pre-trained UNet to a stack of images to generate features for every pixel.
+
+A Worked Example
+-----------------
+
+Let’s start with an example applying a pre-trained [UNet](https://arxiv.org/abs/1505.04597) to a stack of light sheet microscopy data.
+
+In this example, we:
+1. Load the image data from Zarr into a multi-chunked Dask array
+2. Load a pre-trained PyTorch model that featurizes images
+3. Construct a function to apply the model onto each chunk
+4. Apply that function across the Dask array with the dask.array.map_blocks function.
+5. Store the result back into Zarr format
+
+### Step 1. Load the image data
+
+First, we load the image data into a Dask array.
+
+ We were provided sample data from [Prof. Shroff’s lab](https://www.nibib.nih.gov/about-nibib/staff/hari-shroff) at the NIH. The data originally was provided as a 3D TIFF file which we subsequently converted to Zarr with a shape of (950, 2048, 2048). That's a z-stack of 950 images where each slice is 2048x2048 pixels.
+
+This is the same data that we analysed in our last [blogpost on Dask and ITK](https://blog.dask.org/2019/08/09/image-itk). You should note the similarities to that workflow even though we are now using new libaries and performing different analyses.
+
+```
+cd '/Users/nicholassofroniew/Github/image-demos/data/LLSM'
+```
+
+```python
+# Load our data
+import dask.array as da
+imgs = da.from_zarr("AOLLSM_m4_560nm.zarr")
```

Comment:
This data set is actually from Gokul, and not the one above. I could try re-runing on the one above though, would probably be better for consistency reasons

Reply:
Oh, sorry! I saw the sentence "This is the same data that we analysed in our last blogpost on Dask and ITK" so I went ot https://blog.dask.org/2019/08/09/image-itk and copy-pasted the first description of data I saw in that post.

Questions:
1. What description of the data should I include to replace the incorrect bit about "Prof. Shroff’s lab"?
2. Is it correct to say "This is the same data that we analysed in our last blogpost on Dask and ITK"
3. Is the right link to that post https://blog.dask.org/2019/08/09/image-itk

Reply:
> I could try re-runing on the one above though, would probably be better for consistency reasons

I am anti- you re-running this analysis on different data, because I think that might make it more likely to delay things, and I'd like to publish this blogpost on Monday.

Reply:
1. We should say something about it coming from https://science.sciencemag.org/content/360/6386/eaaq1392, used with permission from Srigokul Upadhyayula.

2. Sort of, I think it is the same dataset, but I had my own zarr copy, but that looks like it was also `AOLLSMData_m4` from Gokul.

3. I think that is the right link (or at least I answered 2 based on that)

Reply:
Here is link to data license https://drive.google.com/file/d/1kO9rMRahjaQBV3YUzM8bsd3Noq50q5pX/view?usp=sharing, looks like it is CC-4.0 SA (but you might want to double check)

Reply:
So it seems I got the "Prof Schoff's lab" comment from https://blog.dask.org/2020/11/12/deconvolution. That's **not** the ITK post Nick intended to refer to, the link to the right ITK post is this one (with Gokul's data): https://blog.dask.org/2019/08/09/image-itk

Reply:
> We should say something about it coming from https://science.sciencemag.org/content/360/6386/eaaq1392, used with permission from Srigokul Upadhyayula.

It seems the permanent link to that is: http://dx.doi.org/10.1126/science.aaq1392

Reply:
The data labelled `AO-LLSM` in that article is described in Figure 4 *"Fig. 4. AO-LLSM over large volumes.(A) Aberration-corrected volumerendering over 213mm by 213mm by 113mm in the tail region of a livezebrafish embryo 96 hpf expressing PM-targeted EGFP, assembled fromindependently corrected subvolumes of 7 by 7 by 3 tiles (movie S4)"*

I think that's the right one.

Reply:
I've updated the data attribution.